promql/engine_test.go

```diff
@@ -3256,6 +3259,914 @@ func TestNativeFloatHistogramRate(t *testing.T) {
 	require.Equal(t, expectedHistogram, actualHistogram)
 }
 
+// Tests the rate() function for native histograms with both in-order and OOO sample ingestion and counter resets
+func TestNativeHistogramRateWithCounterResets(t *testing.T) {
+	scenarios := map[string]struct {
+		appendFunc func(appender storage.Appender, lbls labels.Labels, ts, value int64) (storage.SeriesRef, error)
+	}{
+		"integer histogram": {
+			appendFunc: func(appender storage.Appender, lbls labels.Labels, ts, value int64) (storage.SeriesRef, error) {
+				h := tsdbutil.GenerateTestHistogram(int(value))
+				return appender.AppendHistogram(0, lbls, ts, h, nil)
+			},
+		},
+		"float histogram": {
+			appendFunc: func(appender storage.Appender, lbls labels.Labels, ts, value int64) (storage.SeriesRef, error) {
+				fh := tsdbutil.GenerateTestFloatHistogram(int(value))
+				return appender.AppendHistogram(0, lbls, ts, nil, fh)
+			},
+		},
+	}
+	for name, scenario := range scenarios {
+		t.Run(name, func(t *testing.T) {
+			testNativeHistogramRateWithCounterResets(t, scenario.appendFunc)
+		})
+	}
+}
+
+// The ZeroCount, Count, and Sum values in the expectedHistogram come from multiplying
+// the histogram that results from the histogramRate function by a factor. The factor is
+// determined by the formula: (extrapolateToInterval / sampledInterval) / seconds in interval
+func testNativeHistogramRateWithCounterResets(t *testing.T,
+	appendFunc func(appender storage.Appender, lbls labels.Labels, ts, value int64) (storage.SeriesRef, error),
+) {
+	// TODO(beorn7): Integrate histograms into the PromQL testing framework
+	// and write more tests there.
+	seriesName := "sparse_histogram_series"
+	lbls := labels.FromStrings("__name__", seriesName)
+
+	type filterFunc func(v int64) bool
+	defaultFilterFunc := func(v int64) bool { return false }
+
+	type sampleBatch struct {
+		from   int64
+		until  int64
+		filter filterFunc
+	}
+
+	tests := []struct {
+		name              string
+		appendFunc        func(appender storage.Appender, lbls labels.Labels, ts, value int64) error
+		batches           []sampleBatch
+		expectedHistogram *histogram.FloatHistogram
+	}{
+		{
+			name: "In-order samples with counter reset at beginning of interval",
+			batches: []sampleBatch{
+				{
+					from:  0,
+					until: 100,
+					filter: func(v int64) bool {
+						return v == 16 // first sample captured within interval
+					},
+				},
+			},
+			// Since the counter reset is at the first sample in the 5-minute interval
```

Comment:
```suggestion
			// Since the counter reset is at the first sample in the query interval
```

We are only querying samples between 4-5 minutes, so not sure if saying 5-minute interval makes sense